using scheme's equationId function to make sparse matrix structure

adityaghantasala opened the pull request:
As discussed in the issue #192 

Changes are 

- Adding pScheme as an argument to function ResizeAndInitializeVectors  of the base builder and solver.

- Adding pScheme as an argument to ConstructMatrixStructure of the residual based block and elimination builder and solver to use scheme's function EquationId() to formulate the structure of the sparse matrix. (all in the core)

The main idea is to be consistent and facilitate the scheme decide how the sparse matrix looks like (important for implementing Multipoint constraints)

This change requires updating the builder and solvers in the applications.

adityaghantasala committed:
Propagating the pScheme through the functions to reach ConstructMatrixStructure function

jcotela commented:
Aditya, I cannot compile this branch on my machine, It tells me that it cannot find some functions from scheme (`pScheme->EquationId` and `pScheme->Condition_EquationId`). I think we are missing some changes on the scheme class.

adityaghantasala committed:
Propagating the pScheme through the strategies.

adityaghantasala commented:
Hello Jordi 

I forgot to change the strategies and take the pointer of the element in the builder and solver. I pushed the these changes and now it should compile. 

Also when compiling this branch please compile only the core (switch off all the applications). This is because, I did only propagate the pScheme through the core only.

jcotela commented:
Thanks, it is compiling now. I'll check it a bit and come back with my impressions.

jcotela committed:
Avoiding the creation of new pointers

There is no need to create new boost pointers from element and condition iterators, we can retrieve the underlying pointer.

adityaghantasala commented:
Thank you Jordi 

I was also not sure about that part about the pointers. 
Second *(i_condition.base() ... does it give a boost shared pointer ?? or a raw pointer ??!

jcotela commented:
It should give a boost pointer. The `i_element` and `i_condition` are custom kratos iterators that iterate over lists of shared pointers (to elements/conditions) but return references to the element/condition.
Then `i_element.base()` is the standard iterator for the list (of shared pointers) and `*(i_element.base())` returns an object on that list, that is, a boost pointer.

RiccardoRossi commented:
hi guys,

wouldn't it be better to have "ElementEquationId" 
so that the function name looks the same as "ConditionEquationId"?

on a minor remark on the top of Jordi's explanation, if you want to get a raw pointer from a shared pointer you can do it using the get() method of the shared pointer

jcotela commented:
@RiccardoRossi I'm also bothered by the inconsistency in the naming, although I would prefer calling both methods EquationId. Unfortunately, both of them are functions that already exist in the Scheme class, and I guess changing them here would add quite a lot of complexity to an already delicate branch.

RiccardoRossi commented:
@jcotela i agree on calling them EquationId, both of them

I think that the method id New, so It should not be a problem...

adityaghantasala commented:
As @jcotela  suggested changing the names will make it more complex merge.
In the case we want to change the names of the functions the changes should also be propagated through all the core. 

I would prefer to make this as a separate pull request.  

@RiccardoRossi  Thank you for the hint about the  shared pointers. I will keep this in mind. 
@jcotela  @RiccardoRossi  *(i_element.base()) does this work for all the objects like DOFs, schemes etc ??

jcotela commented:
The pointer thing works for anything that is stored internally in a PointerVector or PointerVectorSet container. Elements, conditions and Nodes work like this. DOFs kind of work like this, but they are stored internally as raw pointers (not boost pointers) in some places, so the actual returned object might be a raw pointer sometimes (acutally, I've never checked, but I think so...)

adityaghantasala commented:
@jcotela  Thank you. 

Any hints on how to proceed with the pull request ? 

I am waiting for this so that I can migrate Multipoint constraint implementation to scheme and make it available for others to use it.

jcotela commented:
I think we should start propagating the changes to other applications. I can help with the fluids and trilinos ones, but I don't really know where else this might be used (we should check for custom strategies and builder_and_solvers).

jcotela committed:
Updating trilinos builder and solvers

jcotela committed:
Updating trilinos strategies

jcotela commented:
I commited the required changes in the trilinos_application. The strategies in the fluid dynamics one do not reimplement the ResizeAndInititalize method, so I have not updated them.

jcotela committed:
Adding descriptive error messages

RiccardoRossi commented:
i would suggest that when you touch the files you take care to add the "override" keyword 

this will greatly ease future changes of this type.

actually you may even do it automatically on the top of your whole application. (this is comment is well outside the scope of current conversation, just take it as a suggestion for the future
this is easily done by clang-tidy as:

python /home/rrossi/compiled_libraries/clang-39/share/clang/run-clang-tidy.py -checks=-*,modernize-use-override -fix

(PATH should be modified so that clang-tidy can be found)

RiccardoRossi commented on applications/trilinos_application/python_scripts/trilinos_contact_strategy.py:
```diff
@@ -222,7 +222,7 @@ def InitializeSolutionStep(self, reform_dofs):
             # reorder the list of degrees of freedom to identify fixity and system size
```
i wonder if we should not completely remove this file (as a different pull request of course)

pooyan-dadvand replied:
+1, this is very old and I did it adhoc for tunneling problem with a lot of balancing problem.

RiccardoRossi commented on applications/trilinos_application/python_scripts/trilinos_strategy_python.py:
```diff
@@ -146,7 +146,7 @@ def InitializeSolutionStep(self, reform_dofs):
             # reorder the list of degrees of freedom to identify fixity and system size
```
i wonder if we should not completely remove this file (as a different pull request of course)

RiccardoRossi commented on kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h:
```diff
@@ -1056,9 +1057,10 @@ class ResidualBasedEliminationBuilderAndSolver
+        typename TSchemeType::Pointer pScheme,
 		TSystemMatrixType& A,
 		ElementsContainerType& rElements,
 		ConditionsArrayType& rConditions,
```
shouldn't we have an override here?

jcotela replied:
This function is not overriding anything (builder_and_solver.h does not have a ConstructMatrixStructure method)

RiccardoRossi commented on kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_componentwise.h:
```diff
@@ -347,6 +347,7 @@ class ResidualBasedEliminationBuilderAndSolverComponentwise
+        typename TSchemeType::Pointer pScheme,
         TSystemMatrixPointerType& pA,
         TSystemVectorPointerType& pDx,
         TSystemVectorPointerType& pb,
```
missing overrides

RiccardoRossi reviewed (approved):
i think there are still some missin overrides, however no blocker from my side

+1 for the pull

jcotela committed:
Adding missing override

jcotela commented:
@RiccardoRossi please note that I have only compiled this branch for the applications that I have touched. I can see how this can bring trouble for other applications... @adityaghantasala, have you tried to compile this branch using a more general configure script?

adityaghantasala commented:
@RiccardoRossi  yes as @jcotela  mentioned, this will cause problems to other applications, this is because the strategies and builder and solvers in all other applications should be updated for the change in the base class of the builder and solver. Otherwise the applications will not even compile. 

@jcotela  I did not understand about the configure script. did you mean the configure.sh or any other ? As I mentioned, switching on any other application (except trillinos application, since you modified it) probably will result in errors while compilation.

pooyan-dadvand reviewed (changes requested):
I am just blocking this PR to compile with all applications and run tests for double proofing as suggested @jcotela

RiccardoRossi commented:
ok, i assumed that the other applications were also migrated.

i suggest doing it in two steps: first adding the overrides, and then doing the change

adityaghantasala commented:
@RiccardoRossi  shall I/we change for all the applications ? Or is it possible to inform everyone to make the changes according to the changes to the core ? I will change for structural mechanics applications.

RiccardoRossi commented:
Hi Aditya, i am afraid You should change all the visible applications and
check that their test pass.
I fully understand it is a lot of work, however it is the only way we can
avoid massive code breakage.

Sorry, I would apply the same rule for myself...

jcotela commented:
@adityaghantasala yes, I was thinking about compiling with a more comprehensive configure.sh. Anyway, @pooyan-dadvand is doing this now. We will obtain a list of failing applications to plan the remaining work.

@msandre Could you check if the adjoint application is affected by the changes in this branch? We can not compile the adjoint application here at the moment because we don't have the hdf5 requirements properly set up.

jcotela commented:
@msandre @pooyan-dadvand  I checekd the adjoint application on my computer. It compiles fine when using this branch, but I can't run the tests because the IncompressibleFluidApplication is failing.

msandre commented:
I can try the adjoint app on my machine tomorrow.

adityaghantasala committed:
Merge branch 'master' of https://github.com/KratosMultiphysics/Kratos into scheme_usage_for_sparse_matrix

adityaghantasala committed:
Changing other applications to comply with the changes to the builder and solver

adityaghantasala committed:
Merge branch 'scheme_usage_for_sparse_matrix' of https://github.com/KratosMultiphysics/Kratos into scheme_usage_for_sparse_matrix

Conflicts:
	applications/trilinos_application/custom_strategies/builder_and_solvers/trilinos_block_builder_and_solver.h
	applications/trilinos_application/custom_strategies/builder_and_solvers/trilinos_elimination_builder_and_solver.h

adityaghantasala committed:
Changing the builder and solver of all the applicaitons to make use of the pScheme to make the matrix strucure

adityaghantasala committed:
Minor changes to complete the migration to changed builder and solver function InitializeAndResizeVectors

adityaghantasala commented:
@RiccardoRossi  @pooyan-dadvand @jcotela  

I have migrated and pushed most of the applications with the exception of MultiScaleApplication where there is a different version of the function (which I found confusing).

adityaghantasala commented:
Also I compiled with all the applications turned ON. There was no problem.

RiccardoRossi commented:
regarding Multiscale i believe you are right, the application is non-conformant and should not block the pull request.

@zaghistefano @marandra could you take a look at that? i think the change to bring this up to speed should be on your side and should not block this pull request

jcotela commented:
@adityaghantasala Thanks for taking care of this. Can you tell us the what was the problem with the Multiscale Application? I tried to take a look, but the only builder and solver there was already modified, as far as I can tell.

RiccardoRossi commented:
hi,

i just compiled the branch in windows and i get a segfault in the tests :-(. 

in particular there is a segfault in the test_redistance, which internally uses the builder and solver and the scheme. Does it pass on linux?

adityaghantasala commented:
Hello Riccardo

I can test it. Can you tell me what should I do to exactly run the test you ran ?

jcotela commented:
@RiccardoRossi The core and fluid dynamics tests are passing in my machine. Is it somewhere else?

msandre commented:
AdjointFluidApplication compiles and tests pass on my machine.

RiccardoRossi commented:
@adityaghantasala 
          from kratos/kratos/python_scripts

you can launch all tests by doing
       python3 run_tests.py

specifically the one that fails for me is 
       kratos/kratos/tests/test_redistance.py

@jcotela @msandre i am compiling in windows and in principle i am up to date with the branch, so I fear it is a windows related issue...

adityaghantasala commented:
@jcotela  In MultiscaleApp there is a custom utility "RveLinearSystemOfEquations" which does resizing of the matrices but does not have the same syntax. So I was confused. 

But looking closely, I realized that this is independent of the builder and solver and the changes proposed. So I do not think it is necessary to change it.

jcotela commented:
I confirm that test_redistance.py is failing on windows when using this branch.

RiccardoRossi commented:
@adityaghantasala  i think this shall be reopened no?

adityaghantasala commented:
@RiccardoRossi  yes It should be I by mistake closed it I think. Yes.

jcotela commented:
I'm trying to debug the error in the test that @RiccardoRossi was reporting. I'm catching an exception at the scheme level in `residualbased_incrementalupdate_static_scheme.h` line 291. The exception is seemingly thrown by the element (in `CalculateLocalSystem`?) and says 'Impossible to invert the local matrix for element 15'. The matrix itself (as printed by the error message) is unallocated (its size is 0,0).

Does this tell anything to you, @RiccardoRossi ?

RiccardoRossi commented:
Hi @jcotela taking a peek of the code i may guess we have a inf or nan in some of the distances.
Otherwise this sort of behaviour should not happen. I am not sure if this is the origin of the problem or a consequence of it.

Having said this, i also want to debug this but i was so optimistic to believe that i could reinstall visual studio "fast" and i am still in the installation process...

i'll go back on this asap

RiccardoRossi committed:
fixing the segfault. However this is a risky fix, since it might change the behaviour in some cases. Should be checked carefully

RiccardoRossi committed:
fixing a / t oa \

RiccardoRossi commented:
@jcotela i debugged the error and it appears to be caused by a lack of robustness in the way i handled the case of the distance being exactly 0. I pushed a commit, however this is sort of risky since it might change the existing behaviour in some corner case. @jcotela, could u see what happens with the distance calculation in the case of Eduardo's application

Jordi (Rubio) this might also have an impact in the redistance for your apps (the formation of spurious bubbles). Could you take a peek if this solves your problems?

adityaghantasala commented:
Hello @jcotela  and @RiccardoRossi  any suggestions on how should we further with this pull request ?

RiccardoRossi commented:
Hi Aditya, please hold on still for short time. I will ping again (by telephone this time) a colleague, since i am afraid my change might affect him, and i need his green light. sorry for delaying this...

pooyan-dadvand commented:
Just to add that the changes are ok for me and I'm just waiting for the answer of C2C as @RiccardoRossi mentioned

RiccardoRossi commented:
Hi Aditya, @jordirubio confirmed my fix works. I think you can merge your change.

Sorry for the waiting

RiccardoRossi reviewed (approved):
+1

RiccardoRossi commented:
@pooyan-dadvand do you still want to keep this issue blocked?

pooyan-dadvand commented:
@adityaghantasala If you don't want to develop more in this branch it would be good to delete it

adityaghantasala commented:
@pooyan-dadvand  I think I will delete this branch.
